### PR TITLE
[TC] Build and Run commands for CapsulaHub CLI

### PR DESCRIPTION
**Pull request by IrinaComindaru**
Test cases for CapsulaHub application https://github.com/capsulajs/capsulajs/issues/26

### CODE REVIEW

File: tests/features/capsulahub-application.feature [hunk 1]
```diff
@@ -0,0 +1,80 @@
+
+ Feature: CapsulaHub application
+
+ Background:
+    Given a npm package called @capsulajs/capsula-hub
+    And   a default port 55555
+    And   a valid port is a number between 1 and 65535
```
Comment (MikeSquall) [inline tests/features/capsulahub-application.feature]:
```suggestion
    And   a valid port is a number between 1024 and 65535
```

File: tests/features/capsulahub-application.feature [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+ Scenario: Run Capsula-Hub instance with different config files on different ports
+    When  I run the command `capsula-hub run -l conf.js -p 8888`
+    And   I run the command `capsula-hub run -l conf1.js -p 9999`
+    Then  an workspace instance is created with the config file conf.js and is running on port 8888
+    And   an workspace instance is created with the config file conf1.js and is running on port 9999
+
+ Scenario: Run Capsula-Hub  instance without config file path without specifying the port
+    When I run the command `capsula-hub run -l`
+    Then I expect to receive `error: option `-l, --local <path-to-file>' argument missing`
```
Comment (MikeSquall) [inline tests/features/capsulahub-application.feature]:
```suggestion
    Then I expect to receive `error: option '-l, --local <path-to-file>' argument missing`
```

File: tests/features/capsulahub-application.feature [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ Scenario: Run Capsula-Hub  instance without config file path specifying the port
+    When I run the <command>
+		 |<command>                  |
+		 |capsula-hub run -p 88888   |
+		 |capsula-hub run -l -p 88888|
+    Then I expect to receive  `No flag given to run the app. Please try "capsula-hub run -l <config_file>"`
+
+Scenario: Run Capsula-Hub  instance with valid config file path and with `-p` with no port
+    When I run `capsula-hub run -l conf8888.js -p`
+    Then I expect to receive `error: option `-p, --port <port>' argument missing`
```
Comment (MikeSquall) [inline tests/features/capsulahub-application.feature]:
```suggestion
    Then I expect to receive `error: option '-p, --port <port>' argument missing`
```

**Review by MikeSquall — APPROVED** [review]

File: tests/features/capsulahub-application.feature [hunk 4]
```diff
@@ -0,0 +1,80 @@
+
+ Feature: CapsulaHub application
+
+ Background:
+    Given a npm package called @capsulajs/capsula-hub
+    And   a default port 55555
+    And   a valid port is a number between 1024 and 65535
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: 1024 - 65535 - where did you get this range?

File: tests/features/capsulahub-application.feature [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    And   two existing config files conf.js and conf1.js located in capsulahub/src/config
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: Configuration of the Workspace is an external file and should not exist inside **capsulahub** source code.

File: tests/features/capsulahub-application.feature [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+ Scenario: Run Capsula-Hub  instance with a valid config file path without specifying the port
+    When  I run the command `capsula-hub run -l <path>` in the relevant directory
+		 | <path>         |
+		 | conf.js        |
+		 | config/conf.js |
+		 | src/config/conf.js|
+    Then  the application is running on the default port
+    And   workspace is created with the specified config file
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: As I understand, we will have to go to that localhost to check what has been rendered in browser to check that Workspace has been created with an appropriate config.

File: tests/features/capsulahub-application.feature [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+ Feature: CapsulaHub application
+
+ Background:
+    Given a npm package called @capsulajs/capsula-hub
+    And   a default port 55555
+    And   a valid port is a number between 1024 and 65535
+    And   two existing config files conf.js and conf1.js located in capsulahub/src/config
+
+ Scenario: Run Capsula-Hub  instance with a valid config file path without specifying the port
+    When  I run the command `capsula-hub run -l <path>` in the relevant directory
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: The argument -l will be changed, but it's a minor change - let's wait for the approve of API before changing this in scenario.

File: tests/features/capsulahub-application.feature [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+		 | conf.js        |
+		 | config/conf.js |
+		 | src/config/conf.js|
+    Then  the application is running on the default port
+    And   workspace is created with the specified config file
+
+ Scenario: Run Capsula-Hub instance with a valid config file path specifying a valid port
+    When  I run the command `capsula-hub run -l conf.js -p 44444`
+    Then  the application is running on the specified port
+    And   workspace is created with the specified config file
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: The same - I think we have to open a browser in this port and check

File: tests/features/capsulahub-application.feature [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+		 | 65536|
+		 | 0    |
+    Then  I expect to receive an error
+
+ Scenario: Run Capsula-Hub instance with a config path that does not exist in the repository
+    Given conf3.js config file doesn't exist
+    When  I run the command `capsula-hub run -l conf3.js`
+    Then  I expect to receive `Error: Cannot find module '.../conf3.js'`
+
+ Scenario: Run Capsula-Hub  instance with a wrong path to an existing configuration file
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: Seems redundant. We just have to check, that we will receive an error for invalid path. It does not matter, if the file with the same name exists somewhere else, if we don't specify the correct path to it.

File: tests/features/capsulahub-application.feature [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    When  I run the command `capsula-hub run -l src/conf.js`
+    Then  I expect to receive `Error: Cannot find module '...src/conf.js'`
+
+ Scenario: Run Capsula-Hub  instance with a valid config file path twice on the same port
+    When  I run the command `capsula-hub run -l conf.js -p 55555`
+    And   I run the same command again
+    Then  I expect to receive `Error: listen EADDRINUSE :::55555`
+
+ Scenario: Run Capsula-Hub  instance with 2 valid config file paths on the same port
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: Seems redundant to me. If you try to run an app twice on the same port - there will be the same error. Different configs play no role here.

File: tests/features/capsulahub-application.feature [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+ Scenario: Run Capsula-Hub instance with different config files on different ports
+    When  I run the command `capsula-hub run -l conf.js -p 8888`
+    And   I run the command `capsula-hub run -l conf1.js -p 9999`
+    Then  an workspace instance is created with the config file conf.js and is running on port 8888
+    And   an workspace instance is created with the config file conf1.js and is running on port 9999
+
+ Scenario: Run Capsula-Hub  instance without config file path without specifying the port
+    When I run the command `capsula-hub run -l`
+    Then I expect to receive `error: option '-l, --local <path-to-file>' argument missing`
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: The message might be changed, but the logic is ok.

**Review by dmitriyslaym — COMMENTED** [review]
The majority of tests are still valid, but due to the increased number of args we will have more scenarios.

File: tests/features/capsulahub-application.feature (hunk 4 above)
Comment (VictoriaBordian) [reply tests/features/capsulahub-application.feature]: @MikeSquall suggested this range and I tried to use 65536 port while testing and an error was thrown

File: tests/features/capsulahub-application.feature [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+          |HardcoreServerConfigurationProvider|
+          |HttpServerConfigurationProvider    |
+          |LocalStorageConfigurationProvider  |
+    And   a default path to the output folder "./dist"
+    And   a valid port is a number between 1024 and 65535
+
+  Scenario: Run Capsula-Hub instance with specifying valid arguments
+    Given a valid token with configuration
+    When  I run the command `capsula-hub run --token="token" --configProvider="configProvider" --dispatcherUrl="dispatcherUrl" --port="port"` #add dispatcherUrl only for hardcoreServer provider
+    Then  an workspace is created
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: Not sure that we are able to manually check hardcoreServer now. Or is it possible, @stephanebenayoun ?

File: tests/features/capsulahub-application.feature [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+          |{}        |
+          |{ test: 'test' }|
+          |[]        |
+          |['test']  |
+          |null      |
+          |undefined |
+          |true      |
+          |false     |
+          |0         |
+          |-1        |
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: I think it's not needed - args in a terminal are treated as strings anyway, so one value with smth like "invalidToken" is enough.

File: tests/features/capsulahub-application.feature [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+          |{}        |
+          |{ test: 'test' }|
+          |[]        |
+          |['test']  |
+          |null      |
+          |undefined |
+          |true      |
+          |false     |
+          |-1        |
+    Then  a relevant error is received
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: I think it's not needed - args in a terminal are treated as strings anyway, so one value with smth like "invalidPort" is enough.

File: tests/features/capsulahub-application.feature [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+  Scenario: Run Capsula-Hub instance with an invalid value of <configProvider>
+    Given a valid token with configuration
+    When  I run the command `capsula-hub run --token="token" --configProvider="configProvider" `
+    And   one of the following values for <configProvider> is provided
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: I think it's not needed - args in a terminal are treated as strings anyway, so one value with smth like "invalidProvider" is enough.

File: tests/features/capsulahub-application.feature [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+          |[]        |
+          |['test']  |
+          |null      |
+          |undefined |
+          |true      |
+          |false     |
+          |-1        |
+    Then  a relevant error is received
+
+  Scenario: Run Capsula-Hub instance with an non-existent configProvider throws an error
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: the same as a previous test

File: tests/features/capsulahub-application.feature [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+  Scenario: Run Capsula-Hub instance twice on the same port
+    Given a valid token with configuration
+    And   a valid port
+    When  I run the command `capsula-hub run --token="token" --port="port" `
+    And   I run the same command again
+    Then  a relevant error is received
+
+  Scenario: Run Capsula-Hub  instance with the same token for different configProvider on two different ports
+    Given a valid token with configuration
+    And   two valid ports "port1" and "port2"
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: port should be a number in a string, like "1234" or "8888"

Comment (dmitriyslaym) [reply tests/features/capsulahub-application.feature]: if you use these port1 and port2 as a variable names - it's fine

File: tests/features/capsulahub-application.feature [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+    When  I run the command `capsula-hub run --token="token" --configProvider="hardcoreServer" --dispatcherUrl="dispatcherUrl" --port="port1 "`
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: @stephanebenayoun Can we run hardcoreServer configuration provider?

File: tests/features/capsulahub-application.feature [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+    And   I run the command `capsula-hub run --token="token" --port="port2"`
+    Then  an workspace instance is created that is running on port "port1"
+    And   hardcoreServer is the configuration provider used
+    And   an workspace instance is created that is running on port "port2"
+    And   httpFile is the configuration provider used
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: It makes sense to have a test, where capsulahub runs a few times with different tokens on different ports - to make sure that a different configurations are applied

File: tests/features/capsulahub-application.feature [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+          |{}        |
+          |{ test: 'test' }|
+          |[]        |
+          |['test']  |
+          |null      |
+          |undefined |
+          |true      |
+          |false     |
+          |0         |
+          |-1        |
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: The same problem as for run

File: tests/features/capsulahub-application.feature [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+          | 0        |
+          |{}        |
+          |{ test: 'test' }|
+          |[]        |
+          |['test']  |
+          |null      |
+          |undefined |
+          |true      |
+          |false     |
+          |-1        |
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: The same problem as for run

File: tests/features/capsulahub-application.feature [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: the same problem as for run

**Review by dmitriyslaym — COMMENTED** [review]
capsula-hub command should be replaced by capsulahub

File: tests/features/capsulahub-application.feature (hunk 12 above)
Comment (stephanebenayoun) [reply tests/features/capsulahub-application.feature]: Anyway our tests are running with mocks no ?

Comment (dmitriyslaym) [reply tests/features/capsulahub-application.feature]: I see, we will run with mocks

File: tests/features/capsulahub-application.feature [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+  Background:
+    Given an npm package called @capsulajs/capsula-hub
+    And   a default port 55555
+    And   a default configProvider httpFile
+    And   the following <configProvider> available
+          |<configProvider>                   |
+          |LocalFileConfigurationProvider     |
+          |HttpFileConfigurationProvider      |
+          |HardcoreServerConfigurationProvider|
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: After the latest changes in ConfigurationService we should change **HardcoreServerConfigurationProvider** to **ScalecubeConfigurationService** and **hardcoreServer** to **scalecube**

File: tests/features/capsulahub-application.feature (hunk 12 above)
Comment (dmitriyslaym) [reply tests/features/capsulahub-application.feature]: I think we need this test for each configurationProvider - and the token will be unique for each of providers

File: tests/features/capsulahub-application.feature [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+    Then  an workspace instance with configuration of tenant 1 is created that is running on port "6666" 
+    And   an workspace instance with configuration of tenant 2 is created that is running on port "8888"
+    And   httpFile is the configuration provider used for both workspaces
+    
+  Scenario: Run `capsulahub build` with specifying valid arguments
+    Given a valid token with configuration
+    When  I run the command `capsulahub build --token="token" --configProvider="configProvider" --dispatcherUrl="dispatcherUrl" --output="output"` #add dispatcherUrl only for hardcoreServer provider
+    Then  the app is built
+    And   the result of the build is put in the specified output path
+    And   the specified configProvider is used
```
Comment (dmitriyslaym) [inline tests/features/capsulahub-application.feature]: the same as for run - separate success scenario for each configurationProvider

**Review by dmitriyslaym — APPROVED** [review]

**Pull request merged**